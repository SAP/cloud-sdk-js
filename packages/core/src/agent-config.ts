import * as http from 'http';

/**
 * Interface for the http-agent within the Axios request config.
 */
export interface HttpAgentConfig {
  httpAgent: http.Agent;
}

/**
 * Interface for the https-agent within the Axios request config.
 */
export interface HttpsAgentConfig {
  httpsAgent: http.Agent;
}
