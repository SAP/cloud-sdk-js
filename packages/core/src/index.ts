export * from './scp-cf';
export * from './util';
export * from './odata/common';
export * from './odata/v2';
export * from './odata/v4';
export * from './header-builder';
export * from './http-agent';
export * from './http-client';
export * from './protocol';
export * from './get-protocol';
