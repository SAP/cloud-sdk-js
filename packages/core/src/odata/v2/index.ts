/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

export * from './edm-types';
export * from './entity';
export * from './entity-deserializer';
export * from './entity-serializer';
export * from './filter-functions';
export * from './odata-batch-change-set';
export * from './odata-batch-request-builder';
export * from './odata-batch-retrieve-request';
export * from './payload-value-converter';
export * from './request-builder';
export * from './selectable/custom-field';
export * from './uri-conversion';
export * from './legacy-request-configs';
export * from './extract-odata-etag';
export * from '../common';
