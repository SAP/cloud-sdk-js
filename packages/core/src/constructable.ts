/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { Entity, EntityBuilderType } from './entity';
import { RequestBuilder } from './request-builder/request-builder';
import { CustomField, Selectable } from './selectable';

/**
 * @hidden
 */
export interface Constructable<
  EntityT extends Entity,
  EntityTypeForceMandatoryT = {}
> {
  _serviceName: string;
  _entityName: string;
  _defaultServicePath: string;
  _allFields: Selectable<EntityT>[];
  _keyFields: Selectable<EntityT>[];
  _keys: { [keys: string]: Selectable<EntityT> };
  new (...args: any[]): EntityT;
  requestBuilder(): RequestBuilder<EntityT>;
  builder(): EntityBuilderType<EntityT, EntityTypeForceMandatoryT>;
  customField(fieldName: string): CustomField<EntityT>;
}
