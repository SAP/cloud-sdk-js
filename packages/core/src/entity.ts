/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { MapType } from '@sap-cloud-sdk/util';
import { Constructable, ConstructableODataV4 } from './constructable';
import { EntityBuilder, EntityBuilderODataV4 } from './entity-builder';
import { CustomField, CustomFieldODataV4, Link, LinkODataV4, Selectable, SelectableODataV4 } from './selectable';
import { nonEnumerable } from './util';
import { toPropertyFormat } from './util/name-converter';

export type EntityBuilderType<
  EntityT extends Entity,
  EntityTypeForceMandatoryT
> = {
  [property in keyof EntityTypeForceMandatoryT]: (
    value: EntityTypeForceMandatoryT[property]
  ) => EntityBuilderType<EntityT, EntityTypeForceMandatoryT>;
} &
  EntityBuilder<EntityT, EntityTypeForceMandatoryT>;

export type EntityBuilderTypeODataV4<
  EntityT extends EntityODataV4,
  EntityTypeForceMandatoryT
  > = {
  [property in keyof EntityTypeForceMandatoryT]: (
    value: EntityTypeForceMandatoryT[property]
  ) => EntityBuilderTypeODataV4<EntityT, EntityTypeForceMandatoryT>;
} &
  EntityBuilderODataV4<EntityT, EntityTypeForceMandatoryT>;

// todo fix reference of Entity
export abstract class AbstractEntity {
  static _serviceName: string;
  static _entityName: string;
  static _defaultServicePath: string;

  /**
   * The remote state of the entity.
   * Remote state refers to the last known state of the entity on the remote system from which it has been retrieved or to which it has been posted.
   * It is stored as map, where the keys are stored in the format of the original OData properties.
   */
  protected remoteState: { [keys: string]: any };

  /**
   * The current ETag version of the entity in the remote system.
   * The ETag identified the version of the in the remote system. It will be automatically set in the "if-match" header of update requests and can be set as a custom header for delete requests.
   * When no ETag is provided by the remote system the value of this variable defaults to "*".
   */
  protected _versionIdentifier: string;

  /**
   * A mapper representing custom fields in an entity.
   * Custom fields are represented by their field names and the corresponding values.
   * A custom field can be added or updated using [[setCustomField]] method.
   */
  protected _customFields: MapType<any>;

  constructor() {
    nonEnumerable(this, '_customFields');
    this._customFields = {};
  }

  /**
   * ETag version identifier accessor.
   *
   * @returns The ETag version identifier of the retrieved entity, returns undefined if not retrieved
   */
  get versionIdentifier(): string {
    return this._versionIdentifier;
  }

  /**
   * Returns a map that contains all entity custom fields.
   *
   * @returns A map of all defined custom fields in the entity
   */
  getCustomFields(): MapType<any> {
    return this._customFields;
  }

  /**
   * Custom field value getter.
   *
   * @param fieldName - The name of the custom field
   * @returns The value of the corresponding custom field
   */
  getCustomField(fieldName: string): any {
    return this._customFields[fieldName];
  }

  /**
   * Sets a new custom field in the entity or updates it.
   * Throws an error, if the provided custom field name is already defined by an original field in entity.
   *
   * @param fieldName - The name of the custom field to update
   * @param value - The value of the field
   * @returns The entity itself, to facilitate method chaining
   */
  setCustomField(fieldName: string, value: any): this {
    if (this.isConflictingCustomField(fieldName)) {
      throw new Error(
        `The field name "${fieldName}" is already defined in the entity and cannot be set as custom field.`
      );
    }

    this._customFields[fieldName] = value;

    return this;
  }

  /**
   * Validates whether a custom field exists in the entity.
   *
   * @param fieldName - The name of the custom field to update
   * @returns A boolean value, that indicates whether a custom field is defined in entity
   */
  hasCustomField(fieldName: string): boolean {
    return this._customFields[fieldName] !== undefined;
  }

  /**
   * Sets all retrieved custom fields in entity.
   *
   * @param customFields - Extracted custom fields from a retrieved entity
   * @returns A boolean value, that indicates the existence of the field in entity
   */
  initializeCustomFields(customFields: MapType<any>): this {
    Object.entries(customFields).forEach(cf => {
      this.setCustomField(cf[0], cf[1]);
    });
    return this;
  }

  /**
   * Set the ETag version identifier of the retrieved entity.
   *
   * @param etag - The returned ETag version of the entity
   * @returns The entity itself, to facilitate method chaining
   */
  public setVersionIdentifier(etag: string | undefined): this {
    if (etag && typeof etag === 'string') {
      nonEnumerable(this, '_versionIdentifier');
      this._versionIdentifier = etag;
    }
    return this;
  }

  /**
   * Initializes or sets the remoteState of the entity.
   * This function is called on all read, create and update requests.
   * This function should be called after [[initializeCustomFields]], if custom fields are defined.
   *
   * @deprecated Since 1.12.0. Will be removed in version 2.0.
   * @param state - State to be set as remote state
   * @returns The entity itself, to facilitate method chaining
   */
  public setOrInitializeRemoteState(state?: MapType<any>): this {
    if (!this.remoteState) {
      nonEnumerable(this, 'remoteState');
      this.remoteState = {};
    }
    if (state) {
      Object.entries(state).forEach(([fieldName, value]) => {
        if (this[toPropertyFormat(fieldName)]) {
          this.remoteState[toPropertyFormat(fieldName)] = value;
        } else {
          // We store the custom field with its original name in the remote state
          this.remoteState[fieldName] = value;
        }
      });
    } else {
      this.remoteState = this.getCurrentMapKeys();
    }
    return this;
  }

  /**
   * Returns all updated custom field properties compared to the last known remote state.
   *
   * @returns A map containing all updated custom properties, with their new values
   */
  public getUpdatedCustomFields(): MapType<any> {
    if (this.remoteState === undefined) {
      return this._customFields;
    }

    return Object.entries(this.getCustomFields())
      .filter(([fieldName, value]) => this.remoteState[fieldName] !== value)
      .reduce(
        (updatedCustomFields, [fieldName, value]) => ({
          ...updatedCustomFields,
          [fieldName]: value
        }),
        {}
      );
  }

  /**
   * Returns all changed properties compared to the last known remote state.
   * The returned properties does not include custom fields. Use [[getUpdatedCustomFields]], if updated custom fields are needed.
   *
   * @returns Entity with all properties that changed
   */
  public getUpdatedProperties(): this {
    const current = this.getCurrentMapKeys();
    if (this.remoteState === undefined) {
      return current;
    }
    const patch = {};
    Object.keys(current)
      .filter(key => this.propertyIsEnumerable(key))
      .filter(key => !this.hasCustomField(key))
      .forEach(key => {
        if (this.remoteState[key] !== current[key]) {
          patch[key] = current[key];
        }
      });
    return patch as this;
  }

  /**
   * Returns a map of all defined fields in entity to their current values.
   *
   * @returns Entity with all defined entity fields
   */
  protected getCurrentMapKeys(): this {
    return Object.keys(this)
      .filter(key => this.propertyIsEnumerable(key))
      .reduce(
        (accumulatedMap, key) => ({ ...accumulatedMap, [key]: this[key] }),
        this.getCustomFields()
      ) as this;
  }

  /**
   * Validates whether a field name does not conflict with an original field name and thus can be defined as custom fields.
   *
   * @param customFieldName - Field name to check
   * @returns Boolean value that describes whether a field name can be defined as custom field
   */
  protected isConflictingCustomField(customFieldName: string): boolean {
    return this[toPropertyFormat(customFieldName)] !== undefined;
  }
}

interface ODataV2Version{
  oDataV2Version: true;
}

interface ODataV4Version{
  oDataV4Version: true;
}

/**
 * Super class for all representations of OData entity types.
 */
export class Entity extends AbstractEntity implements ODataV2Version{
  oDataV2Version: true = true;

  protected static entityBuilder<
    EntityT extends Entity,
    EntityTypeForceMandatoryT
    >(
    entityConstructor: Constructable<EntityT, EntityTypeForceMandatoryT>
  ): EntityBuilderType<EntityT, EntityTypeForceMandatoryT> {
    const builder = new EntityBuilder<EntityT, EntityTypeForceMandatoryT>(
      entityConstructor
    );
    entityConstructor._allFields.forEach(field => {
      const fieldName = `${toPropertyFormat(field._fieldName)}`;
      builder[fieldName] = function (value) {
        this.entity[fieldName] = value;
        return this;
      };
    });
    return builder as EntityBuilderType<EntityT, EntityTypeForceMandatoryT>;
  }

  protected static customFieldSelector<EntityT extends Entity>(
    fieldName: string,
    entityConstructor: Constructable<EntityT>
  ): CustomField<EntityT> {
    return new CustomField(fieldName, entityConstructor);
  }
}

export class EntityODataV4 extends AbstractEntity implements ODataV4Version{
  oDataV4Version: true = true;

  protected static entityBuilder<
    EntityT extends EntityODataV4,
    EntityTypeForceMandatoryT
    >(
    entityConstructor: ConstructableODataV4<EntityT, EntityTypeForceMandatoryT>
  ): EntityBuilderTypeODataV4<EntityT, EntityTypeForceMandatoryT> {
    const builder = new EntityBuilderODataV4<EntityT, EntityTypeForceMandatoryT>(
      entityConstructor
    );
    entityConstructor._allFields.forEach(field => {
      const fieldName = `${toPropertyFormat(field._fieldName)}`;
      builder[fieldName] = function (value) {
        this.entity[fieldName] = value;
        return this;
      };
    });
    return builder as EntityBuilderTypeODataV4<EntityT, EntityTypeForceMandatoryT>;
  }

  protected static customFieldSelector<EntityT extends EntityODataV4>(
    fieldName: string,
    entityConstructor: ConstructableODataV4<EntityT>
  ): CustomFieldODataV4<EntityT> {
    return new CustomFieldODataV4(fieldName, entityConstructor);
  }
}

/**
 * @hidden
 */
export interface EntityIdentifiable<T extends Entity> {
  readonly _entityConstructor: Constructable<T>;
  readonly _entity: T;
}

export interface EntityIdentifiableODataV4<T extends EntityODataV4> {
  readonly _entityConstructor: ConstructableODataV4<T>;
  readonly _entity: T;
}

/* eslint-disable valid-jsdoc */

/**
 * @hidden
 */
export function isSelectedProperty<EntityT extends Entity>(
  json,
  selectable: Selectable<EntityT>
) {
  return json.hasOwnProperty(selectable._fieldName);
}

export function isSelectedPropertyODataV4<EntityT extends EntityODataV4>(
  json,
  selectable: SelectableODataV4<EntityT>
) {
  return json.hasOwnProperty(selectable._fieldName);
}

/**
 * @hidden
 */
export function isExistentProperty<
  EntityT extends Entity,
  LinkedEntityT extends Entity
>(json, link: Link<EntityT, LinkedEntityT>) {
  return isSelectedProperty(json, link) && json[link._fieldName] !== null;
}

export function isExistentPropertyODataV4<
  EntityT extends EntityODataV4,
  LinkedEntityT extends EntityODataV4
  >(json, link: LinkODataV4<EntityT, LinkedEntityT>) {
  return isSelectedPropertyODataV4(json, link) && json[link._fieldName] !== null;
}

/**
 * @hidden
 */
export function isExpandedProperty<
  EntityT extends Entity,
  LinkedEntityT extends Entity
>(json, link: Link<EntityT, LinkedEntityT>) {
  return (
    isExistentProperty(json, link) &&
    !json[link._fieldName].hasOwnProperty('__deferred')
  );
}

export function isExpandedPropertyODataV4<
  EntityT extends EntityODataV4,
  LinkedEntityT extends EntityODataV4
  >(json, link: LinkODataV4<EntityT, LinkedEntityT>) {
  return (
    isExistentPropertyODataV4(json, link) &&
    !json[link._fieldName].hasOwnProperty('__deferred')
  );
}
