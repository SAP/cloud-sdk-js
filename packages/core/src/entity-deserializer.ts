/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { MapType } from '@sap-cloud-sdk/util';
import { Constructable } from './constructable';
import { Entity, isExpandedProperty, isSelectedProperty } from './entity';
import { edmToTs } from './payload-value-converter';
import {
  CollectionField,
  ComplexTypeField,
  EdmTypeField,
  Link,
  OneToOneLink,
  Selectable
} from './selectable';
import { toPropertyFormat } from './util';

/**
 * Extracts all custom fields from the JSON payload for a single entity.
 * In this context, a custom fields is every property that is not known in the corresponding entity class.
 *
 * @param json - The JSON payload.
 * @param entityConstructor - The constructor function of the entity class.
 * @returns An object containing the custom fields as key-value pairs.
 */
export function extractCustomFields<EntityT extends Entity, JsonT>(
  json: Partial<JsonT>,
  entityConstructor: Constructable<EntityT>
): MapType<any> {
  const regularODataProperties = [
    '__metadata',
    '__deferred',
    ...entityConstructor._allFields.map(field => field._fieldName)
  ];
  const regularFields = new Set<string>(regularODataProperties);
  return Object.keys(json)
    .filter(key => !regularFields.has(key))
    .reduce((customFields, key) => {
      customFields[key] = json[key];
      return customFields;
    }, {});
}

/**
 * Converts the JSON payload for a single entity into an instance of the corresponding generated entity class.
 * It sets the remote state to the data provided by the JSON payload.
 * If a version identifier is found in the '__metadata' or in the request header, the method also sets it.
 *
 * @param json - The JSON payload.
 * @param entityConstructor - The constructor function of the entity class.
 * @param requestHeader - Optional parameter which may be used to add a version identifier (etag) to the entity
 * @returns An instance of the entity class.
 */
export function deserializeEntity<EntityT extends Entity, JsonT>(
  json: Partial<JsonT>,
  entityConstructor: Constructable<EntityT>,
  requestHeader?: any
): EntityT {
  const etag = extractODataETag(json) || extractEtagFromHeader(requestHeader);
  return entityConstructor._allFields
    .filter(field => isSelectedProperty(json, field))
    .reduce((entity, staticField) => {
      entity[toPropertyFormat(staticField._fieldName)] = getFieldValue(
        json,
        staticField
      );
      return entity;
    }, new entityConstructor())
    .initializeCustomFields(extractCustomFields(json, entityConstructor))
    .setVersionIdentifier(etag)
    .setOrInitializeRemoteState();
}

export function deserializeEntityODataV4<EntityT extends Entity, JsonT>(
  json: Partial<JsonT>,
  entityConstructor: Constructable<EntityT>,
  requestHeader?: any
): EntityT {
  const etag = extractODataETag(json) || extractEtagFromHeader(requestHeader);
  return entityConstructor._allFields
    .filter(field => isSelectedProperty(json, field))
    .reduce((entity, staticField) => {
      entity[toPropertyFormat(staticField._fieldName)] = getFieldValueODataV4(
        json,
        staticField
      );
      return entity;
    }, new entityConstructor())
    .initializeCustomFields(extractCustomFields(json, entityConstructor))
    .setVersionIdentifier(etag)
    .setOrInitializeRemoteState();
}

function extractEtagFromHeader(headers: any): string | undefined {
  return headers ? headers['Etag'] || headers['etag'] : undefined;
}

function extractODataETag(json: MapType<any>): string | undefined {
  return '__metadata' in json ? json['__metadata']['etag'] : undefined;
}

function getFieldValue<EntityT extends Entity, JsonT>(
  json: Partial<JsonT>,
  selectable: Selectable<EntityT>
) {
  if (selectable instanceof EdmTypeField) {
    return edmToTs(json[selectable._fieldName], selectable.edmType);
  } else if (selectable instanceof Link) {
    return getLinkFromJson(json, selectable);
  } else if (selectable instanceof ComplexTypeField) {
    return deserializeComplexType(json[selectable._fieldName], selectable);
  }
}

function isODataV2Selectable<EntityT extends Entity>(
  selectable: Selectable<EntityT>
): boolean {
  return (
    selectable instanceof EdmTypeField ||
    selectable instanceof Link ||
    selectable instanceof ComplexTypeField
  );
}

function getFieldValueODataV4<EntityT extends Entity, JsonT>(
  json: Partial<JsonT>,
  selectable: Selectable<EntityT>
) {
  if (isODataV2Selectable(selectable)) {
    return getFieldValue(json, selectable);
  } else if (selectable instanceof CollectionField) {
    return deserializeCollectionType(json[selectable._fieldName], selectable);
  }
}

function getLinkFromJson<
  EntityT extends Entity,
  LinkedEntityT extends Entity,
  JsonT
>(json: Partial<JsonT>, link: Link<EntityT, LinkedEntityT>) {
  return link instanceof OneToOneLink
    ? getSingleLinkFromJson(json, link)
    : getMultiLinkFromJson(json, link);
}

// Be careful: if the return type is changed to `LinkedEntityT | undefined`, the test 'navigation properties should never be undefined' of the 'business-partner.spec.ts' will fail.
// Not sure the purpose of the usage of null.
function getSingleLinkFromJson<
  EntityT extends Entity,
  LinkedEntityT extends Entity,
  JsonT
>(
  json: Partial<JsonT>,
  link: Link<EntityT, LinkedEntityT>
): LinkedEntityT | null {
  if (isExpandedProperty(json, link)) {
    const entity = deserializeEntity(json[link._fieldName], link._linkedEntity);
    return entity;
  }
  return null;
}

function getMultiLinkFromJson<
  EntityT extends Entity,
  LinkedEntityT extends Entity,
  JsonT
>(
  json: Partial<JsonT>,
  link: Link<EntityT, LinkedEntityT>
): LinkedEntityT[] | undefined {
  if (isSelectedProperty(json, link)) {
    const results = json[link._fieldName].results || [];
    return results.map(linkJson =>
      deserializeEntity(linkJson, link._linkedEntity)
    );
  }
}

function deserializeComplexType<EntityT extends Entity>(
  json: MapType<any>,
  selectable: ComplexTypeField<EntityT>
): MapType<any> {
  return Object.entries(selectable)
    .filter(
      ([_, field]) =>
        field instanceof EdmTypeField &&
        typeof json[field._fieldName] !== 'undefined'
    )
    .reduce((complexTypeObject, [fieldName, field]) => {
      complexTypeObject[toPropertyFormat(fieldName)] = edmToTs(
        json[field._fieldName],
        field.edmType
      );
      return complexTypeObject;
    }, {});
}

function deserializeCollectionType<EntityT extends Entity>(
  json: any[],
  selectable: CollectionField<EntityT>
) {
  if (selectable._elementType instanceof EdmTypeField) {
    const edmType = selectable._elementType.edmType;
    return json.map(v => edmToTs(v, edmType));
  } else if (selectable._elementType instanceof ComplexTypeField) {
    const complexTypeField = selectable._elementType;
    return json.map(v => deserializeComplexType(v, complexTypeField));
  }
}
