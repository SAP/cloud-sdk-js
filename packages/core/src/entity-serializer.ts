/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { MapType } from '@sap-cloud-sdk/util';
import { Constructable } from './constructable';
import { Entity } from './entity';
import { tsToEdm } from './payload-value-converter';
import {
  CollectionField,
  ComplexTypeField,
  EdmTypeField,
  Link,
  OneToOneLink
} from './selectable';
import { toStaticPropertyFormat } from './util';

/**
 * Converts an instance of an entity class into a JSON payload to be sent to an OData service.
 *
 * @param entity - An instance of an entity.
 * @param entityConstructor - The constructor function of that entity.
 * @returns JSON.
 */
export function serializeEntity<EntityT extends Entity>(
  entity: EntityT,
  entityConstructor: Constructable<EntityT>
): MapType<any> {
  return {
    ...serializeEntityNonCustomFields(entity, entityConstructor),
    ...entity.getCustomFields()
  };
}

export function serializeEntityODataV4<EntityT extends Entity>(
  entity: EntityT,
  entityConstructor: Constructable<EntityT>
): MapType<any> {
  return {
    ...serializeEntityNonCustomFieldsODataV4(entity, entityConstructor),
    ...entity.getCustomFields()
  };
}

/**
 * Converts an instance of an entity class into a JSON payload to be sent to an OData service, ignoring custom fields.
 *
 * @param entity - An instance of an entity.
 * @param entityConstructor - The constructor function of that entity.
 * @returns JSON.
 */
export function serializeEntityNonCustomFields<EntityT extends Entity>(
  entity: EntityT,
  entityConstructor: Constructable<EntityT>
): MapType<any> {
  if (!entity) {
    return {};
  }
  return Object.keys(entity).reduce((serialized, key) => {
    const selectable = entityConstructor[toStaticPropertyFormat(key)];
    const fieldValue = entity[key];
    if (isODataV2Serizable(fieldValue, selectable)) {
      serialized[selectable._fieldName] = serializeField(
        fieldValue,
        selectable
      );
    }

    return serialized;
  }, {});
}

export function serializeEntityNonCustomFieldsODataV4<EntityT extends Entity>(
  entity: EntityT,
  entityConstructor: Constructable<EntityT>
): MapType<any> {
  if (!entity) {
    return {};
  }
  return Object.keys(entity).reduce((serialized, key) => {
    const selectable = entityConstructor[toStaticPropertyFormat(key)];
    const fieldValue = entity[key];

    if (isODataV2Serizable(fieldValue, selectable)) {
      serialized[selectable._fieldName] = serializeField(
        fieldValue,
        selectable
      );
    } else if (selectable instanceof CollectionField) {
      serialized[selectable._fieldName] = serilizeCollectionField(
        fieldValue,
        selectable
      );
    }

    return serialized;
  }, {});
}

function serilizeCollectionField<EntityT extends Entity>(
  fieldValue: any[],
  selectable: CollectionField<EntityT>
) {
  if (selectable._elementType instanceof EdmTypeField) {
    const edmType = selectable._elementType.edmType;
    return fieldValue.map(v => tsToEdm(v, edmType));
  } else if (selectable._elementType instanceof ComplexTypeField) {
    const complexTypeField = selectable._elementType;
    return fieldValue.map(v => serializeComplexTypeField(complexTypeField, v));
  }
}

function isODataV2Serizable(fieldValue, selectable) {
  return (
    fieldValue === null ||
    fieldValue === undefined ||
    selectable instanceof EdmTypeField ||
    selectable instanceof OneToOneLink ||
    selectable instanceof Link ||
    selectable instanceof ComplexTypeField
  );
}

function serializeField(fieldValue, selectable) {
  if (fieldValue === null || fieldValue === undefined) {
    return null;
  } else if (selectable instanceof EdmTypeField) {
    return tsToEdm(fieldValue, selectable.edmType);
  } else if (selectable instanceof OneToOneLink) {
    return serializeEntityNonCustomFields(fieldValue, selectable._linkedEntity);
  } else if (selectable instanceof Link) {
    return fieldValue.map(linkedEntity =>
      serializeEntityNonCustomFields(linkedEntity, selectable._linkedEntity)
    );
  } else if (selectable instanceof ComplexTypeField) {
    return serializeComplexTypeField(selectable, fieldValue);
  }
}

function serializeComplexTypeField<EntityT extends Entity>(
  selectable: ComplexTypeField<EntityT>,
  fieldValue: any
): any {
  return Object.entries(selectable).reduce(
    (complexTypeObject, [propertyKey, propertyValue]) => {
      const value = fieldValue[propertyKey];
      if (
        propertyValue instanceof EdmTypeField &&
        typeof value !== 'undefined'
      ) {
        complexTypeObject[propertyValue._fieldName] = tsToEdm(
          fieldValue[propertyKey],
          propertyValue.edmType
        );
      }
      return complexTypeObject;
    },
    {}
  );
}
