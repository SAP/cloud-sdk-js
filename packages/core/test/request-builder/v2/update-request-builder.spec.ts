/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
import nock from 'nock';
import { v4 as uuid } from 'uuid';
import { UpdateRequestBuilderV2 } from '../../../src';
import { muteLoggers } from '../../test-util/mute-logger';
import {
  defaultDestination,
  mockUpdateRequest
} from '../../test-util/request-mocker';
import { testEntityResourcePath } from '../../test-util/test-data';
import { TestEntity } from '../../test-util/test-services/v2/test-service';

function createTestEntity() {
  const keyPropGuid = uuid();
  const keyPropString = 'stringId';
  const int32Prop = 125;

  return TestEntity.builder()
    .keyPropertyGuid(keyPropGuid)
    .keyPropertyString(keyPropString)
    .int32Property(int32Prop)
    .build();
}

describe('UpdateRequestBuilderV2', () => {
  beforeAll(() => {
    muteLoggers('http-agent');
  });

  afterEach(() => {
    nock.cleanAll();
  });

  it('no request is executed when there is no difference between current and remote state', async () => {
    const entity = createTestEntity().setOrInitializeRemoteState();
    const scope = nock(/.*/).get(/.*/).reply(500);
    const actual = await new UpdateRequestBuilderV2(TestEntity, entity).execute(
      defaultDestination
    );
    expect(actual).toEqual(entity);
    expect(scope.isDone()).toBe(false);
  });

  it('no request is executed when only keys have been modified', async () => {
    const scope = nock(/.*/).get(/.*/).reply(500);
    const entity = createTestEntity().setOrInitializeRemoteState();
    entity.keyPropertyGuid = uuid();
    entity.keyPropertyString = 'UPDATED!';
    const actual = await new UpdateRequestBuilderV2(TestEntity, entity).execute(
      defaultDestination
    );
    expect(actual).toEqual(entity);
    expect(scope.isDone()).toBe(false);
  });

  it('update request sends only non-key properties', async () => {
    const entity = createTestEntity();
    entity.booleanProperty = false;
    const requestBody = {
      Int32Property: entity.int32Property,
      BooleanProperty: false
    };

    mockUpdateRequest({
      body: requestBody,
      path: testEntityResourcePath(
        entity.keyPropertyGuid,
        entity.keyPropertyString
      )
    });

    const actual = await new UpdateRequestBuilderV2(TestEntity, entity).execute(
      defaultDestination
    );
    expect(actual).toEqual(entity.setOrInitializeRemoteState());
  });

  it('update request includes new custom fields', async () => {
    const entity = createTestEntity();
    const customFieldVal = 'CUSTOM';
    entity.setCustomField('SomeCustomField', customFieldVal);

    const requestBody = {
      Int32Property: entity.int32Property,
      SomeCustomField: customFieldVal
    };

    mockUpdateRequest({
      body: requestBody,
      path: testEntityResourcePath(
        entity.keyPropertyGuid,
        entity.keyPropertyString
      )
    });

    const actual = await new UpdateRequestBuilderV2(TestEntity, entity).execute(
      defaultDestination
    );
    expect(actual).toEqual(entity.setOrInitializeRemoteState());
  });

  it('update request includes updated custom fields', async () => {
    const entity = createTestEntity()
      .setCustomField('SomeCustomField', 'CUSTOM')
      .setOrInitializeRemoteState();

    const customFieldVal = 'UPDATED!';
    entity.setCustomField('SomeCustomField', customFieldVal);

    const requestBody = { SomeCustomField: customFieldVal };

    mockUpdateRequest({
      body: requestBody,
      path: testEntityResourcePath(
        entity.keyPropertyGuid,
        entity.keyPropertyString
      )
    });

    const actual = await new UpdateRequestBuilderV2(TestEntity, entity).execute(
      defaultDestination
    );
    expect(actual).toEqual(entity.setOrInitializeRemoteState());
  });

  it('update request sends the whole entity when using PUT', async () => {
    const entity = createTestEntity();

    const putRequestBody = {
      KeyPropertyGuid: entity.keyPropertyGuid,
      KeyPropertyString: entity.keyPropertyString,
      Int32Property: entity.int32Property
    };

    mockUpdateRequest({
      body: putRequestBody,
      path: testEntityResourcePath(
        entity.keyPropertyGuid,
        entity.keyPropertyString
      ),
      method: 'put'
    });

    const actual = await new UpdateRequestBuilderV2(TestEntity, entity)
      .replaceWholeEntityWithPut()
      .execute(defaultDestination);

    expect(await actual).toEqual(entity.setOrInitializeRemoteState());
  });

  it('update request includes required fields', async () => {
    const entity = createTestEntity().setOrInitializeRemoteState();
    const requestBody = { KeyPropertyGuid: entity.keyPropertyGuid };

    const scope = mockUpdateRequest({
      body: requestBody,
      path: testEntityResourcePath(
        entity.keyPropertyGuid,
        entity.keyPropertyString
      )
    });

    const actual = await new UpdateRequestBuilderV2(TestEntity, entity)
      .requiredFields(TestEntity.KEY_PROPERTY_GUID)
      .execute(defaultDestination);

    expect(scope.isDone()).toBe(true);
    expect(actual).toEqual(entity);
  });

  it('update request excludes ignored properties', async () => {
    const entity = createTestEntity();

    const scope = nock(/.*/).patch(/.*/).reply(500);

    const actual = await new UpdateRequestBuilderV2(TestEntity, entity)
      .ignoredFields(TestEntity.INT_32_PROPERTY)
      .execute(defaultDestination);

    expect(scope.isDone()).toBe(false);
    expect(actual).toEqual(entity);
  });

  it('update request should contain version identifier when set on entity', async () => {
    const entity = createTestEntity().setVersionIdentifier('not-a-star');
    const requestBody = { Int32Property: entity.int32Property };

    mockUpdateRequest({
      body: requestBody,
      path: testEntityResourcePath(
        entity.keyPropertyGuid,
        entity.keyPropertyString
      ),
      additionalHeaders: { 'if-match': 'not-a-star' }
    });

    const actual = await new UpdateRequestBuilderV2(TestEntity, entity).execute(
      defaultDestination
    );
    expect(actual).toEqual(entity.setOrInitializeRemoteState());
  });

  it('update request should contain version identifier when set on request', async () => {
    const entity = createTestEntity().setVersionIdentifier('not-a-star');
    const requestBody = { Int32Property: entity.int32Property };
    const customVersionIdentifier = 'custom-version-identifier';

    mockUpdateRequest({
      body: requestBody,
      path: testEntityResourcePath(
        entity.keyPropertyGuid,
        entity.keyPropertyString
      ),
      additionalHeaders: { 'if-match': customVersionIdentifier }
    });

    const actual = await new UpdateRequestBuilderV2(TestEntity, entity)
      .withCustomVersionIdentifier(customVersionIdentifier)
      .execute(defaultDestination);

    expect(actual).toEqual(entity.setOrInitializeRemoteState());
    expect(actual.versionIdentifier).toBe(customVersionIdentifier);
  });

  it('should ignore version identifier', async () => {
    const entity = createTestEntity().setVersionIdentifier('not-a-star');
    const requestBody = { Int32Property: entity.int32Property };

    mockUpdateRequest({
      body: requestBody,
      path: testEntityResourcePath(
        entity.keyPropertyGuid,
        entity.keyPropertyString
      ),
      additionalHeaders: { 'if-match': '*' }
    });

    const actual = await new UpdateRequestBuilderV2(TestEntity, entity)
      .ignoreVersionIdentifier()
      .execute(defaultDestination);

    expect(actual).toEqual(entity.setOrInitializeRemoteState());
  });

  it('throws an error when request execution fails', async () => {
    const entity = createTestEntity();

    mockUpdateRequest({
      body: () => true,
      path: testEntityResourcePath(
        entity.keyPropertyGuid,
        entity.keyPropertyString
      ),
      statusCode: 500
    });

    const updateRequest = new UpdateRequestBuilderV2(
      TestEntity,
      entity
    ).execute(defaultDestination);

    await expect(updateRequest).rejects.toThrowErrorMatchingSnapshot();
  });
});
