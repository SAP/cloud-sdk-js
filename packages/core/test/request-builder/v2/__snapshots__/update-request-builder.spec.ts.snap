// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateRequestBuilderV2 throws an error when request execution fails 1`] = `"OData update request failed!"`;

exports[`UpdateRequestBuilderV2 warns if navigaton properties are send 1`] = `"OData update request failed!"`;
